Blocks: Support block nesting (Add Columns block, InnerBlocks component)

aduth opened the pull request:
Closes #2995 
Closes #428
Partially implements #219
~Cherry-picks 1efbb9faf02b12a64759fddc7027e5bfa089ecfc and 2bd32f8cf0acb0dcbf183f8edae348b32ab1e9b3 from #3743~
Related: #3705, #2743
~Partially reverts #2806 (temporarily)~

This pull request seeks to explore support for generalized nested blocks, implementing a `InnerBlocks` component which effectively behaves as a sub-editor block listing (maintaining its own separate editor state). A block's children are available via its `block.innerBlocks` property, and can be accessed from a component's `edit` and `save` implementations (`props.innerBlocks`), and updated via `edit`'s `setInnerBlocks` prop.

Included is a new Columns block which behaves similarly to the Text Columns block, except that it allows insertion of blocks within the rendered columns.

![Nesting](https://user-images.githubusercontent.com/1779930/33456167-8f3efe96-d5ec-11e7-8177-5304c436ec3b.png)

**What has been implemented:**

- Block parsing, creation, validation, and serialization of nested content
- Editing block children, including insertion, rearrangement, toolbar and inspector controls

**What doesn't work so well yet:**

- Front-end display
- Reconciliation of selected blocks, particularly in rendering of toolbar and inspector slot contents
- Reducing columns trims content instead of collapsing
- General polish and UX behaviors
- Tests and such

__Testing instructions:__

Verify that you can insert a new Columns block, and within it, insert, rearrange, edit, and delete child blocks. Verify that saving the post and refreshing the editor respects all changes which were made.

aduth committed:
Editor: Move slot provider to top-level

Nesting editors still want to make use of ancestored slots, e.g. inserter, toolbar, inspector controls

aduth committed:
Block API: Implement support for parsing, generating, serializing nested blocks

aduth committed:
Block List: Enable block edit to revise inner blocks

aduth committed:
Blocks: Add InnerBlocks component for nested editor

aduth committed:
Blocks: Add Columns block

georgestephanis commented:
Summing up a brief discussion we had in person, general wish-list and considerations (in no way is any of this a blocker or need-to-have for initial merge, just musings I wanted to get enumerated somewhere)

Another example (apart from layout use-cases) would be Contact Forms.  Some bits for the future to support that use case:

* Some blocks would not make sense in a non-nested context -- for example, Fields would not make sense unless you're adding them to a form.  So registering blocks, but flagging them as disabled unless in a certain parent?  Or adding them to a blacklist on the general context, to still allow them elsewhere?  Unsure on implementation.
* It would be useful to be able to define a default template for nested blocks (as per #3668) -- so for example, a contact form instantiation could start with Name, Email Address, and Message fields every time.  (or a Recipe block could start with an Image, List of Ingredients, and Directions blocks)

There were other things not immediately coming to mind, I'll add them as they come up.

aduth commented:
>Some blocks would not make sense in a non-nested context -- for example, Fields would not make sense unless you're adding them to a form. So registering blocks, but flagging them as disabled unless in a certain parent? Or adding them to a blacklist on the general context, to still allow them elsewhere? Unsure on implementation.

Noting that it wasn't specifically needed for a columns block (which should allow any child type), but I had an in-progress change to be able to support this via a `block.parents` array:

```diff
diff --git a/editor/components/inserter/menu.js b/editor/components/inserter/menu.js
index 73bfb7c0..4dc344fb 100644
--- a/editor/components/inserter/menu.js
+++ b/editor/components/inserter/menu.js
@@ -126,6 +126,12 @@ export class InserterMenu extends Component {
                                return false;
                        }
 
+                       // If block defines parent nesting, disallow except when explicitly
+                       // provided as a whitelisted block type.
+                       if ( Array.isArray( block.parents ) ) {
+                               return includes( blockTypes, block.name );
+                       }
+
                        // Block types defined as either `true` or array:
                        //  - True: Allow
                        //  - Array: Check block name within whitelist
@@ -311,7 +317,7 @@ export class InserterMenu extends Component {
                                }
                                { isSearching &&
                                        <div role="menu" className="editor-inserter__search-results">
-                                               { this.renderCategories( this.getVisibleBlocksByCategory( getBlockTypes() ) ) }
+                                               { this.renderCategories( this.getVisibleBlocksByCategory( this.getBlockTypes() ) ) }
                                        </div>
                                }
                        </TabbableContainer>
```

The thinking being a block could define it's allowed parents for nesting context, and conversely, a parent would define supportable children via a prop on `BlockChildren` (`types={ [ 'core/text', ... ] }`).

aduth commented:
>It would be useful to be able to define a default template for nested blocks (as per #3668) -- so for example, a contact form instantiation could start with Name, Email Address, and Message fields every time. (or a Recipe block could start with an Image, List of Ingredients, and Directions blocks)

I think this would be a good idea: Perhaps as an `initialValue` or `defaultValue` on `BlockChildren`, or as part of the block API to define default children. There's also a general question as to whether children needs to be a separate value of a block, or can be an attribute as any other (in which case leveraging attributes `default` behavior).

mtias commented on blocks/api/parser.js:
```diff
@@ -132,12 +132,22 @@ export function getBlockAttributes( blockType, innerHTML, attributes ) {
+	attributes = attributes || {};
+
+	// Trim content to avoid creation of intermediary freeform segments
+	innerHTML = innerHTML.trim();
```
How was it being created otherwise?

mtias replied:
I see it was already present just in a different moment.

mtias commented on blocks/api/serializer.js:
```diff
@@ -32,16 +32,17 @@ export function getBlockDefaultClassname( blockName ) {
  * @return {string}            Save content
  */
-export function getSaveContent( blockType, attributes ) {
+export function getSaveContent( blockType, attributes, children = [] ) {
```
What's the general thinking behind sometimes referring to `innerBlocks` and sometimes referring to `children`?

aduth replied:
>What's the general thinking behind sometimes referring to `innerBlocks` and sometimes referring to `children`?

We should probably align it. `innerBlocks` is the naming provided from the parser, but `children` seemed a friendlier / more familiar name. Maybe too familiar though, since it conflicts with the React built-in prop of the same name; could mislead a developer to think they could render it as an element set (which it is not).

Do you have a preference one way or the other?

mcsf replied:
I'd lean toward `innerBlocks` for the reasons you cite, plus the fact that with `innerBlocks` we reinforce the idea of "blocks" as the central object. Also, the notion of containment behind _inner_ may be more fitting than that of descent behind _children_.

mtias commented on blocks/api/serializer.js:
```diff
@@ -62,7 +63,12 @@ export function getSaveContent( blockType, attributes ) {
+	let content = renderToString( contentWithExtraProps );
+
+	// Drop raw HTML wrappers (support dangerous inner HTML without wrapper)
+	content = content.replace( /<\/?wp-raw-html>/g, '' );
```
Is it better to drop after rendering to string? What can be included in `wp-raw-html`?

aduth replied:
>Is it better to drop after rendering to string? What can be included in `wp-raw-html`?

This is a hacky workaround to the [lack of `<React.Fragment dangerouslySetInnerHTML />` support](https://wordpress.slack.com/archives/C02QB2JS7/p1512663464000044) in React. It achieves the desired result by rendering a wrapper with a known name which is intended to be dropped during serialization. It would be used in cases where we want to render raw markup without a wrapper element (e.g. here for nested blocks, in #3048 for the saved output of Editable).

aduth commented:
In rebasing this, I've found that our recent changes to treat the editor as a singleton store (#3832) conflict with the intent here to introduce nested block editing via separate store instances. @youknowriad do you have any suggestions on how I might proceed here? Maybe I just need to `createStore` with a direct reference to the reducer? I'll plan to spend some more time on this tomorrow.

youknowriad commented:
> @youknowriad do you have any suggestions on how I might proceed here? Maybe I just need to createStore with a direct reference to the reducer?

Yes, I think using "Provider" and "it just works" was hiding the fact that the "blocks" become "local state" with the changes introduce in the current PR. 

I think we should have a separate store for blocks (not a singleton) instantiated in the `BlockChildren` component. Ideally it would be used in the root level as well. and the main "store" would only contain a "blocks" reducer maybe being updated with the `onChange` of the `BlockChildren` component.

I've previously explored the idea in https://github.com/WordPress/gutenberg/pull/2065

aduth commented:
>I think we should have a separate **store for blocks** (not a singleton) instantiated in the `BlockChildren` component. Ideally it would be used in the root level as well. and the main "store" would only contain a **"blocks" reducer** maybe being updated with the `onChange` of the `BlockChildren` component.

What do you have in mind being managed in each of these cases? A primary benefit of creating the nested provider context is not needing to consider implications of hierarchy and nesting in the store itself, which is quite convenient. It's also not just blocks, but state such as selection, hover, etc.

youknowriad commented:
@aduth 

So the root store would handle "post", "layout" (since edit-post is still there), "saving", "metaboxes", "editor", "edits"...

The blocks store would be "blocks", "selection", "hover" (all what's block related)

Do we need some of these informations in the upper level? maybe, I don't know right now, this is discoverable while implementing.

---

This is not different from what you were doing originally. It's just making it clear that the "BlocksProvider" is about the blocks only since I assume the other parts of the store were useless in the "BlockChildren", am I wrong? 

I believe a quicker alternative to mimic exactly what you were doing is to recreate the store, expose a `createStore` from "editor/store/index.js" and use it in `BlockChildren`. Maybe I'm wrong but I think the separate stores is nicer, it clarifies what handles what.

jorgefilipecosta commented:
Hi @aduth, I have been testing this branch and generally, it works well, nice work 👍  There are some things to polish but I think this is a good shape and we may solve polish/solve some bugs.

Regarding the singleton store vs multiple stores discussion, while the multiple stores being used may make things simple at the start, I think it may cause us some problems in the future to make some features works as expected. E.g: undo may affect the content of multiple stores.

I think we may be able to use a single store. We can have a higher order reducer "H" that would apply to a reducer that has "multiple instances" "S" e.g: editor, blockSelection. We would dispatch an action to create an instance e.g:"{type: CREATE_EDITOR_CONTEXT, id: uuid}. The reducer "H" would create the instance for us e.g: assign to an object the reducer "S" passed to it to generate the empty state.
The actions affecting the reducer "S" would need to contain the uuid and our higher order reducer "H" would call the reducer with the right state branch. If no uuid is passed our higher order reducer "H" calls our reducer "S" on all branches.
The selectors selecting branches "S" would receive only the branch instead of the state, there would exist a selector to select the right branch based on the uuid. 
With this approach, we would be able to apply actions to all the "contexts" e.g: remove all multi-selections, and we would retain the advantages of a single store, e.g: being able to debug with redux dev tools.

youknowriad commented:
Just a small note that the separate store approach (which is just a convenient way to build a complex local state for a react component) brings us closer to the "Block Editor" being a generic JS module reusable in several contexts (inside WordPress aka Customizer for instance and outside Wordpress).

mcsf commented:
With regards to nesting proper—ignoring the weaknesses cited in the description—this seems to work well.

`BlockChildren` is quite interesting. I like the idea that we can swap contexts at that membrane and let its element subtree behave without necessary knowledge of its own nesting. _A priori_ this seems more viable than juggling enhanced reducers, although I used to feel more strongly about striving for a single store.

aduth commented:
Force pushed a rebased copy of the branch, bringing the changes up to date, reconciling with the revised state approach, and being more consistent with the naming of "inner blocks" vs. "children" (deferring to the former, since the latter is overloaded already).

I'm observing some strange behaviors with toolbar interactions, which might be related to the block selection not working correctly anymore in the nested block (should be deselecting parent). Going to see if recent revisions in #3743 improve this.

Needs a plan for front-end styling, and for tests to be updated.

aduth commented:
Noting another issue is that trying to "Show Advanced Settings" from a block in the visual editor will not display the sidebar. This is because since the nested block component maintains its own full copy of the editor store, it's also expected to manage the display of sidebars (which it doesn't). Toward @youknowriad's earlier points, we likely need to break this down further where the only state managed by `InnerBlocks` is that of the blocks it's showing.

youknowriad commented:
Some notes:

 - Is it intentional to make the columns block "wide" or just a leftover?
 - The inspector shows only the container options, how do we show the inner block inspector
 - The columns block doesn't have a block toolbar, what behavior should it have if it had one (I assume showing the inner block's toolbar unless the "container" is selected)
 - I tried multi-selecting some children blocks and clicking backspace. It doesn't work

If we want to make this perfect before merging, it will take a lot of time. So we might want to make a decision here. Do we want to merge this (even a bit broken sometimes) and iterate on it (which could lead to a lot of issues created once released), or should we iterate on the PR which could lead to a giant PR.

mtias commented:
I am fine with merging an initial implementation if we mark the "columns block" with some text that says it is experimental, and provided we don't break other areas of the editor.

aduth committed:
Revert "Editor: Move slot provider to top-level"

This reverts commit 7b6adced07889fe8eaf2b769f7708b0fc58033aa.

JasonHoffmann commented:
Not sure if this is exactly the right place for this, but I wanted to hop in and just highlight another use case I've been running into. Being able to embed / nest one block inside of another programmatically in a more structured way.

For instance, let's say I have a grid of items. Each one has an image, a title, and then some "paragraph" text. What I want to be able to do is register a custom block that allows users to select how many columns in this grid they want to add, and then for each one, nest an "image" a "heading" and a "paragraph" block inside of a larger column container. Right now (I think) I'd have to recreate the functionality for each custom block, but it would be great just to do something like `addEditInnerBlock( 'core/paragraph' );`

I'm not sure if this means creating an `InnerBlock` component that takes the name of a block as a property, or if there's even a way to do it now. But I wanted to flag it because I think being able to make pre-structured, nested blocks will be a pretty common issue for developers.

aduth commented:
>I'm not sure if this means creating an `InnerBlock` component that takes the name of a block as a property,

Yes, this is how I imagine this working. A block author will be able to support nesting in their own blocks, limited to a subset of specific blocks if desired. Similarly, it might be desirable to allow blocks to opt to only be insertable within specific contexts.

aduth commented:
Given some of the issues highlighted in previous comments, I've been iterating over the past week to move away from the nested provider context toward handling nested blocks in a single store. I've made some good progress and am hoping to propose the changes here this week.

aaronjorbin commented:
Love the concept and where this is going. I would love to see the documentation updates included in one of the near-term updates to that PR.  From my perspective, the doc updates can help with the DUX review a lot more than reading code. (also, this PR is awesome).  Also Unit Tests.

aduth committed:
Framework: Add store and BlockList support for nested context

aduth commented:
I've pushed the latest revised approach, which includes migrating away from a nested editor provider and multiple controls toward including nested block awareness in the single editor store. It does this by allowing `BlockList` to be a reusable component leveraged by the public-facing `InnerBlocks` component, accepting a `rootUID` prop of the block from which nesting is expected to occur.

Further, I have eliminated all responsibility of a block implementer to introspect or segment inner blocks, instead enforcing inner blocks as a single continuous flow of markup. In order to create the effect of distinct columns in the Columns block, `BlockList` has also been enhanced to include understanding of a `layouts` configuration. The idea of layouts is described in exhaustive detail in the [included documentation for `InnerBlocks`](https://github.com/WordPress/gutenberg/tree/add/block-children/blocks/inner-blocks/README.md). While not implemented here, the expectation is that this could pair with changes proposed in #4314 to allow top-level layout configuration for wide and full width appearance, in place of current behaviors to implement using `getEditWrapperProps`.

The circular dependency between `InnerBlocks` and `BlockList` is no more, and has been satisfied instead with context-provided `BlockList` via partial application of a `rootUID` prop provided by the ancestor `BlockList`. Similarly, `InnerBlocks.Content` is will have its content provided via context at the serialization step via a new [`BlockContentProvider` component](https://github.com/WordPress/gutenberg/tree/add/block-children/blocks/block-content-provider/README.md).

The result of all this is that `InnerBlocks` is ideally very simple for block implementers to use. See the [columns block implementation](https://github.com/WordPress/gutenberg/tree/add/block-children/blocks/library/columns/) as a reference example resource, including its [output markup](https://github.com/WordPress/gutenberg/blob/add/block-children/blocks/test/fixtures/core__columns.serialized.html). In subsequent pull requests this will be enhanced to support additional restrictions on the types of blocks which can be nested.

A major challenge in the implementation here was ensuring that blocks do not attempt to handle the events of nested content. The result of this is the [`IgnoreNestedEvents` component](https://github.com/WordPress/gutenberg/tree/add/block-children/editor/components/block-list/ignore-nested-events.js), which is essentially a glorified `stopEventPropagation` without actually stopping  event propagation. This should handle both DOM-based and virtual event bubbling.

__Current Status:__

*User-Facing Behaviors:* For the most part, all behaviors should be implemented and working reasonably well to preserve current block ordering and selection, insert a columns block, change number of columns, and add blocks within columns. A remaining expected behavior is to allow users to move blocks between columns, which as documented in the `InnerBlocks` component README.md is not much more than changing its layout position.

*Block-Author API:* `InnerBlocks` should but does not yet support restrictions on block types allowed within a block. We should probably warn authors who attempt to assign a conflicting `layout` attribute name to their block; we already have one example of this in our [Latest Posts block](https://github.com/WordPress/gutenberg/blob/d967017d528040df234dde882c2ea2419203d55d/blocks/library/latest-posts/index.php#L81-L84).

*Tests:* I made my way through updating a large number of tests taking into account the changes made here, and also stubbing out test cases for testing newly-introduced nested context handling. This effort is labor-intensive but a non-concern.

*Code:* I'm largely happy with the direction here. As it impacts the implementation, I have a few further ideas for refactoring, but none of which is blocking:

 - Refactor insertion point to behave not on index / layout, but on before / after a specific UID, which is less complex to handle in nested context.
 - Decide if there is a better way to structure state to support pointers between levels of nesting (e.g. previous block, parent block). I had briefly worked on a [linked list](https://en.wikipedia.org/wiki/Linked_list) style structure before abandoning it, but something like this may serve better if previous / next / parent is a common operation.

Some of the repetition of `rootUID` and `layout` is quite unfortunate, but seems a natural consequence of including support for nested content.

jasmussen commented:
Amazing work, Andrew. From just a quick glance, the most recent implementation seems to work far better than the previous implementation. 🔥 

As a sidenote, it feels like now that child blocks are going to be more prevalent, improvements to the side UI is going to become more important. I'm working on that, and have a bunch of UI ideas, including https://github.com/WordPress/gutenberg/pull/4539, but those should be able to benefit this UI later on.

jaswrks commented:
Impressive work @aduth! 🎉 

I just took this branch for a spin and noticed two things:

- Using this _exact_ sequence of actions, there's a loss of one paragraph on refresh, along with a loss of the column layout. Tested in latest Chrome on macOS.

![2018-01-21_02-01-46](https://user-images.githubusercontent.com/1563559/35193519-564cfe76-fe50-11e7-976f-89ee46137d0d.gif)

- If a column's paragraph block is completely empty (no text), I'm unable to paste anything into it. Typing a single letter, _then pasting_, works fine.

jaswrks commented on editor/components/block-list/index.js:
```diff
@@ -3,294 +3,44 @@
-			</div>
+			<BlockListLayout
+				key={ layout.name || 'default' }
+				layout={ layout.name }
```
@aduth This is just a minor thing you may already be aware of.

![2018-01-21_02-21-03](https://user-images.githubusercontent.com/1563559/35193628-d36850d0-fe51-11e7-80bc-c63f4a75c910.png)

marcato15 commented:
I noticed the same issue where the parent block was being lost. I tried debugging and noticed that the save function is not being triggered for parent blocks if they have children blocks (even though the children blocks are being saved). I was hoping to actually find the root problem but lost the trail after that.

BE-Webdesign commented:
Seems like a good step forward. Thank you for tackling this issue, it is very appreciated.

I think I am screwing this up somehow, as I am getting an array for my columns block? It appears to be the columns count, and the second item is how many child blocks there are:

```
<!-- wp:columns {"columns":[2,5]} -->
<div class="wp-block-columns">
    <!-- wp:paragraph -->
    <p>Abezzzz</p>
    <!-- /wp:paragraph -->

    <!-- wp:image -->
    <figure class="wp-block-image"><img/></figure>
    <!-- /wp:image -->
    <!-- wp:list -->
    <ul>
        <li>List list list</li>
        <li>Lis</li>
        <li>a</li>
    </ul>
    <!-- /wp:list -->

    <!-- wp:paragraph -->
    <p></p>
    <!-- /wp:paragraph -->

    <!-- wp:cover-image {"url":"http://local.wordpress.dev/wp-content/uploads/2016/03/test2.jpg","id":37} -->
    <section class="wp-block-cover-image has-background-dim" style="background-image:url(http://local.wordpress.dev/wp-content/uploads/2016/03/test2.jpg)">
        <h2>Taco</h2>
    </section>
    <!-- /wp:cover-image -->
</div>
<!-- /wp:columns -->
```

I am also unable to delete the block or access any of the options from the ellipses icon on the top right of the block. Moving it works great!

Is there supposed to be a way to have the columns block have knowledge of the structure of the inner blocks? Right now it appears to be tied to the way it is parsing. The inner blocks is just a list of the inner blocks with no structure of ordering how is the columns block figuring out which blocks go in which column?

I can also not adjust any attributes of the inner blocks like alignment bold/italic etc.

coderkevin commented:
This doesn't seem to be saving correctly for me either. When I save with columns, they aren't preserved in the resulting markup. See the capture below:

![child-blocks-columns](https://user-images.githubusercontent.com/945228/35358444-ae60cd28-0124-11e8-9a18-f52a461da49e.gif)

aduth commented:
>I think I am screwing this up somehow, as I am getting an array for my columns block? It appears to be the columns count, and the second item is how many child blocks there are:

The serialized attributes look like that of the previous iteration. Wondering if you're either running an older version of the branch, or if you had updated a post containing an old columns block?

>This doesn't seem to be saving correctly for me either. When I save with columns, they aren't preserved in the resulting markup.

This may be linked to the related blocking pull requests #4005 and/or #4591.

I'll plan to revisit this again soon to polish the remaining issues.

aduth committed:
Block List: Move BlockList to layout file

Aiding in GitHub rebase history as explicit move

aduth commented:
There were in fact some issues with generating markup for nested block contents, but they turned out to not be the result of the changes here, but in fact a bug in the underlying selector memoization library, with a fix at https://github.com/aduth/rememo/commit/aab2fd3f09c5cc4998ed40c3381094cc76b9a4c8 pending release in the next day or so.

aduth committed:
Block: Stop propagation but don't handle child events

aduth commented:
References to a few commits which will be dropped in rebase as redundant with #4539, which has been referenced as experimental and potentially revertable:

- 1549bfece35abef9455a95014c3f192c06f57408 (reverted in 0ce3efa5e7cab10411511ca529a226a9f474663e)
- 737006a3cbfa8a6294bb427baf6c30a58ce61d44

aduth committed:
Blocks: Rename latest posts layout attribute as postLayout

Otherwise conflicting with newly-introduced "layout" attribute applying to all blocks for nesting arrangement

aduth commented:
This pull request is now ready for final review.

A few major issues have been noted as a result of this exploration, though not directly related to the proposed changes themselves:

- Block migrations assume that a block must be invalid in order to be migrated. This is problematic here because I had wanted to migrate current Latest Posts blocks' `layout` attribute which otherwise conflicts with the newly-introduced global layout attribute applied for all blocks. Since the block is technically valid, the migration is not run.
   - https://gist.github.com/aduth/abcf59040198e28fd8c177230de849d3
- Our strategies for generated class name (and now layout class name) work fine for static blocks, but do not work at all for dynamic blocks. Consider the [Latest Posts block rendering](https://github.com/WordPress/gutenberg/blob/49f05e53a4e9ceff0c08fe646f3b0bb724f58b53/blocks/library/latest-posts/index.php#L50) which manually applies the class name `wp-block-latest-posts` which would otherwise be automatically generated for static blocks. The same applies for custom class name and now layout. While a block implementer could similarly reimplement these behaviors in their block render function, this is fragile and unaccommodating with more default behaviors added in the future.
   - The same issue affects static blocks which return a string from `save`, though we may want to discontinue support for this anyways, especially now with the changes proposed here we have a `<DangerousHTML />` component which can serve in its place.

Some arguably unrelated changes were included in the latest rebase relevant for sibling inserter changes (#4539), which have also been extracted to a separate pull request #4765 in case we want to merge it separately, at which time I can rebase this branch to reduce the number of changes.

jasmussen commented:
This feels _solid_ in a way previous versions did not do.

Just due to the complexity of this PR, and the amount of work it entailed, and how important it is, I feel like that is enough for us to merge this PR as fast as we can.

Some of the things we'll want to fix post-merge, are all around the UI:

- the focus indicator when typing but outlines have faded out is a bit aggressive
- multi selection styles are a bit off when selecting across parent blocks
- multi selection when invoked via the keyboard fails after selecting more than a single nested block

But again — the surrounding stuff feels impressive. Amazing work. 🏅

youknowriad commented on blocks/api/serializer.js:
```diff
@@ -31,20 +32,21 @@ export function getBlockDefaultClassname( blockName ) {
 
 	let saveElement;
 
 	if ( save.prototype instanceof Component ) {
```
Unrelated to this PR: Is there any valid reason to still support "Component" in the save method? It creates the illusion that you can use lifecycle and state but you can't in reality.

mtias replied:
Would be good to follow up on this separately.

aduth replied:
One reason might be reusability of a same component between `edit` and `save`, though this sounds unlikely. I'd argue we shouldn't necessarily _prevent_ the use of a component, but similarly it would be nice if we could not target it explicitly; which is to say, the problem here isn't that we're checking `Component`, it's that below we call it as a function. In both the class and function component, I think we should be using `createElement`.

youknowriad replied:
Why should we treat it as a React Component when we don't support it being a React Component. It's just a function in all cases?

youknowriad commented on blocks/inner-blocks/index.js:
```diff
@@ -0,0 +1,18 @@
+InnerBlocks = withContext( 'BlockList' )()( InnerBlocks );
+
+InnerBlocks.Content = ( { BlockContent } ) => {
+	return <BlockContent />;
```
Why did you choose to rely on the context instead of making it explicit, passing the `innerBlocks` and using them here? Since I assume both `edit` and `save` receive those? I have a personal preference for explicitness but I'm probably missing some context here.

aduth replied:
I see two reasons:

- Because if we pass the inner blocks explicitly, it might give the illusion that it's a controlled input, meaning we'd similarly want to handle `onChange`, which we (currently) have no capability of doing.
   - That said, `<BlockList onChange={ ... } />` doesn't sound like an awful idea in general, and we could "fake" it with a noop and context until it's implementable
- The more pertinent point (one which I've not argued strongly for, but I know @mtias is advocating) is the desire for an isolation between a parent and its nested blocks. As implemented here a block has no need or ability to introspect or manage its nested content.
- Minor point: It would require additional props `innerBlocks` and `setInnerBlocks`.
   - These existed in earlier commits of the branch.

youknowriad replied:
Fine enough, but we can't still make it explicit without the need to pass the inner blocks (which would assume an ònChange`), we could just pass the block `uid` which is also available as a prop I think.

youknowriad replied:
Anyway, it's not that important, just wanted to get rid of this Provider because it's not easy to follow where and how it's applied. I guess it's a common pitfall for all context based components.

youknowriad commented on blocks/api/serializer.js:
```diff
@@ -165,11 +172,14 @@ export function getBlockContent( block ) {
+	// still occur.
 	let saveContent = block.originalContent;
-	if ( block.isValid ) {
+	if ( block.isValid || block.innerBlocks.length ) {
```
So, If I understand properly we're ignoring the validation because we can't validate properly the container blocks? Can you explain a bit more why we can't validate the container blocks properly? (I'm certain you already did elsewhere, just a link maybe :))

Why are we creating a difference between blocks with inner blocks and blocks without inner blocks. If we can't guarantee the behavior for all blocks, shouldn't we just remove the condition to make it consistent?

aduth replied:
The main issue is that if we mark the block as invalid, it will defer the next save to using its `innerHTML`; this is usually meant to preserve the original content, but in a case of a block which contains nested blocks, it destroys the content, because nested content is not reflected in a block's `innerHTML`.

mtias replied:
Also, that we'd be validating the same content n-times.

youknowriad commented on blocks/hooks/layout.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+addFilter( 'blocks.registerBlockType', 'core/anchor/attribute', addAttribute );
+addFilter( 'blocks.getSaveContent.extraProps', 'core/anchor/save-props', addSaveProps );
```
I guess the name of the hooks should be changed here?

aduth replied:
>I guess the name of the hooks should be changed here?

Fixed in rebased 680029e.

youknowriad commented on blocks/inner-blocks/README.md:
```diff
@@ -0,0 +1,147 @@
+
+### `InnerBlocks`
+
+#### `layouts`
```
Maybe I'm wrong (actually, I'm probably wrong but worth asking :P), so this feels like a way to do two dimension handling of inner blocks, which is IMO specific to the columns blocks. Can't we avoid this entirely and use a private inner block `core/column` for each column. So the columns block would be a list of `core/column` and each `core/column` would contain another list of blocks? What's not possible with this one dimension approach?

mtias replied:
The thinking is the extra dimension is semantically unnecessary.

BE-Webdesign replied:
I would be in favor of the private `core/column` block, in my experience it makes handling the document smoother. I also don't understand how this is all coming together, so I am probably not the best candidate for judgement.

aduth commented:
Noted a couple issues with slash inserter and creating reusable blocks in a nested context. Seems to generally be an issue of preserving the value of a globally-applied attribute, in this case the new `layout` attribute. Thus far I've updated the cases manually (the `cloneBlock` API was added to assist), but wondering if there's a better way to handle this: maybe filter for cases where we're creating a new block using another as a base.

mtias commented:
Unrelated, but we have an issue for duplicating a block so `cloneBlock` might be useful regardless.

GlennMartin1 commented:
Is this on for 2.2?  ETA?

coderkevin commented:
Looks like columns still aren't saving correctly for me. 🙁

![child-blocks-columns](https://user-images.githubusercontent.com/945228/35842756-12200728-0ab9-11e8-9f63-5e70ad1558c5.gif)

GlennMartin1 commented:
@coderkevin The present "columns" block is experimental, very buggy, and questionable if it has a future in Gutenberg.

This "nesting" issue is a much greater infrastructure allowing blocks inside blocks, which will make columns possible with a range of other blocks within the columns.

However up to 2.1, there is no nesting. I expect (?) 2.2 to have this nesting merged; that can't be soon enough.

aduth commented:
I've pushed a rebased branch resolving latest conflicts and bringing in changes for `RawHTML` (previously `DangerousHTML`) merged separately in #4786. This also resolves a previous bug where unselecting a block by clicking the editor background was not well.

This should work fairly well, save for a couple remaining issues:

- Multi-selection shift expanding behavior differs from master, where shift-selecting upwards multiple times and then shift-selecting downward will reverse the selection direction from the originating block.
   - In master, it will collapse the expansion toward the originating block. This is arguably the correct behavior, but is made difficult by nesting because concepts of "previous" and "next" block require traversing nested hierarchy.
- A few remaining transforms on nested blocks within Columns will cause the block to disappear (e.g. "Turn into List").
   - This is because the original block's `layout` attribute needs to be preserved through the transformation. I've addressed this in a few cases, but in a very ad-hoc fashion. I think this needs to be addressed at a framework level, perhaps in the `switchToBlockType` block factory function. Since this is at least an issue isolated to Columns, it may be fine to address separately.

>@coderkevin The present "columns" block is experimental, very buggy, and questionable if it has a future in Gutenberg.

@GlennMartin1 The Columns block is technically experimental, yes, but mostly functional. I'd expect the issue raised by @coderkevin to be working correctly, and have asked him to give it another try with a fresh copy of the branch after the latest rebases.

mtias commented:
This is working very well for me. Great job!

One thing to fix is that the grid elements are centered on the front-end.

![image](https://user-images.githubusercontent.com/548849/35875709-6d29e780-0b70-11e8-8d92-68c22a2087f3.png)

The other would be to add the wide/full-width options to the columns block container, as it's what opens up so many possibilities.

aduth commented:
@mtias Is it possible the centering is a style applied by your theme? I'd see no reason why it would be appearing as centered, and does not for me in Twenty Seventeen.

![Left-aligned](https://user-images.githubusercontent.com/1779930/35875959-e0a13880-0b3e-11e8-883a-7abc089a2e15.png)

lumberman commented:
@aduth can you please add 5 columns option as well?

Now we can have any possible number of columns and variations but 5 columns option.
I know my clients will need it for pricing tables.

aduth committed:
Blocks: Add wide / full align options

aduth commented:
>@aduth can you please add 5 columns option as well?

There were already styles for 2-6 columns, so should have allowed these to be assigned. Fixed in 0c1e6f4. 

>The other would be to add the wide/full-width options to the columns block container

Added in a6ee8da.

aduth committed:
Layout: Preserve layout attribute via transform hook

mtias commented:
@aduth yes, it's a margin `auto` applied to some elements. Seems fine to disregard.

mtias commented:
> The reason I might have hesitated is that we may want to deprecate wide alignment in favor of wide layout, with the new layouting system implemented here.

Yes, but we might apply that transformation across the board based on https://github.com/WordPress/gutenberg/pull/4314

mtias commented:
@jasmussen this is for later, but I think we should expose column count control in the toolbar (not only inspector) as it seems important enough to be a main interaction point.

aduth committed:
Blocks: Add experimental modifier to Columns block title

paulwilde commented:
For columns I think one consideration would be to allow for difference widths to apply per column. So say you had 2 columns, having the ability to drag from in between each column to make a 1/3 - 2/3 split. This would allow for having columns where you had an image in the smaller column with accompanying text in the larger one.

Another idea would be to add a blacklist (or whitelist) to set which blocks can be applied to each column width split. It doesn't make sense having the ability to having 3 cover image blocks side-by-side as an example, as it puts too much burden on developers to ensure every single configuration looks right for each block type.

Obviously these are desired features on top of the block nesting implementation. I'll create separate tickets about both once this gets merged.

aduth commented:
@paulwilde Both valid ideas, and both of which appear to be quite compatible with what's currently been implemented here. I'd agree it'd be good to pursue these enhancements separately.

aduth commented:
While I fully expect some rough edges remain, I believe this is finally ready for merge as an initial implementation. Despite the large number of changes, there were actually many pull requests which were split off from this one as prerequisite blockers, noted below for future reference:

#4005
#4585
#4591
#4684
#4707
#4708
#4732
#4765
#4786

**Major Changes**

- `BlockList` enhanced to support two additional props:
   - `rootUID`: Either `''` (empty string) for top-level blocks, or the UID of the parent block from which nested blocks are to descend
      - Many selectors and action creators have been updated to require that a `rootUID` is specified as an argument
   - `layouts`: [See related documentation](https://github.com/WordPress/gutenberg/blob/31a2ce0f3786fc029e17505cb10ce5af6bb848da/blocks/inner-blocks/README.md#layouts)
- New `InnerBlocks` and `InnerBlocks.Content` components which can be used by block implementers to edit and save nested blocks, respectively
   - [See documentation](https://github.com/WordPress/gutenberg/blob/31a2ce0f3786fc029e17505cb10ce5af6bb848da/blocks/inner-blocks/README.md)
- New (experimental) Columns block
   - Note: This block is implemented using [CSS grid styling](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout) ([browser support](https://caniuse.com/#feat=css-grid))
- New block API: `wp.blocks.cloneBlock`
- Updated block API: `wp.block.createBlock` now accepts a third argument, `innerBlocks`
- New filter: `blocks.switchToBlockType.transformedBlock` for filtering an individual transformed result of `switchToBlockType`
- New "global" attribute `layout`: All blocks support this attribute, though it is currently only ever assigned in the Columns block
   - Subsequent pull requests will explore leveraging the `layout` attribute as a replacement for the current "wide" and "full" alignments
- Changed block: The "Posts List" block `layout` attribute has been updated to `postsLayout` per the addition of the new global `layout` attribute (**breaking change**)
- State selector changes
   - New selectors: `getBlockRootUID`, `getAdjacentBlock`
   - Renamed selectors: `getBlockUids` -> `getBlockOrder`
   - Removed selectors: `isFirstBlock`, `isLastBlock`

**Known Bugs**

Multi-select shift expansion behaves differently when collapsing a selection (see https://github.com/WordPress/gutenberg/pull/3745#issuecomment-363496670)

jasmussen commented:
🔥🎉

carlhancock commented:
Why the insistence on defaulting the columns to having Paragraph blocks in them instead of simply having an insert block UI within the column as the default?

By defaulting to paragraph blocks already being in place it means people are constantly going to be deleting the default block after placing the actual blocks they want to use in the columns.

To make it even more annoying... if they don't do things in the correct order when doing what i've described above the process is going to be frustrating.  For example if I add a Columns block to my page and then I edit a column and say I don't want a paragraph here I want an image... so I delete the Paragraph block that is there by default.  It then adds another default Paragraph block.  You have to add the block you actually want to use below that paragraph block and then delete the paragraph block.

I think the assumption that everything defaults to a Paragraph is a mistake.  There are going to be all kinds of blocks created by Gutenberg and the insistence on defaulting to Paragraph is a very blog-centric way of looking at things.  It would be a lot more helpful to look at things from a WordPress as a CMS standpoint.  

It would be a much better user experience if columns allowed you to start from scratch and didn't make the assumption that the first block should be a Paragraph and instead simply provided an insert block UI so you could insert only the blocks you want to insert and not have to go back and clean up what Gutenberg has done for you by default.

gubbigubbi commented:
Great work team. 

> New (experimental) Columns block
> Note: This block is implemented using CSS grid styling (browser support)

CSS grid is the future, but should we look to use flexbox for now, it has much wider browser support? prefixed Flexbox works as far back as IE10.

Eventually we will also need the following features:

- Mobile / tablet / desktop columns
- Padding and margins for each column

aduth commented:
>Why the insistence on defaulting the columns to having Paragraph blocks in them instead of simply having an insert block UI within the column as the default?

Where was this insisted? The current behavior of the Columns block is implemented as it is largely as a consequence of it inheriting behaviors of the underlying `BlockList` components it reuses. In other words, the block is something of an embedded editor. As an initial exploration, the UX of the  block is very unfinished, and I would expect there to be some iterations more toward what you're describing with layout editing.

jaswrks commented on blocks/inner-blocks/README.md:
```diff
@@ -0,0 +1,147 @@
+
+Because `InnerBlocks.Content` will generate a single continuous flow of block markup for nested content, it may be advisable to use [CSS Grid Layout](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout) to assign layout positions. Be aware that CSS grid is [not suported in legacy browsers](https://caniuse.com/#feat=css-grid), and therefore you should consider how this might impact yo ...
+
+Layouts can be assigned either either as an object (ungrouped layouts) or an array of objects (grouped layouts). These are documented below.
```
Layouts can be assigned `either either` → `either`

jasmussen commented:
> Why the insistence on defaulting the columns to having Paragraph blocks in them instead of simply having an insert block UI within the column as the default?

I responded in more detail to [a similar question you asked here](https://github.com/WordPress/gutenberg/pull/4539#issuecomment-363679501). 

We already have insertion UI in the editor bar, and on the side, and using the slash command for the power users. These all work in a nested context, but require your caret to be where you want the content to be inserted, similar to most other editors that let you insert non-text content, like Word or Pages or Google Docs. 

When #3627 gets addressed, the concern of "why is there already a paragraph block here" becomes moot, as any content you insert while the caret is an empty paragraph block _replaces_ it.

yaplusplus commented:
Hey, thanks for the great work! Are there any plans to implement multiple `InnerBlocks` instances inside a nested block? We're making our own WYSIWYG editor and we've developed a special wrapper block to insert the editor into a Gutenberg post. Now we're thinking how to make our editor contain another nested Gutenberg blocks inside its content (note: our editor uses it's own content structure that differs from Gutenberg's).

We use code like that for running the editor:

```js
import SetkaEditor from './SetkaEditor';

export default class SetkaEditorBlock extends Component {

    componentDidMount() {
        // Run editor at the specified DOM node
        SetkaEditor.start(this._domElement, this.props.attributes.content);
    }

    render() {
        // Render wrapper block
        return (
            <div>
                <div className="stk-editor" ref={el => this._domElement = el} />
            </div>
        )
    }

}
```

And probably the final HTML code containing both our and Gutenberg blocks should look like this:
```html
<!-- wp:setka-editor/setka-editor -->

<div class="stk-post">

  <p>Our editor's text paragraph</p>

  <!-- wp:block-1 -->
  <p>Gutenberg block</p>
  <!-- /wp:block-1 -->

  <div>
    <h2>Our editor's heading</h2>

    <!-- wp:block-2 -->
    <div>Another Gutenberg block<div>
    <!-- /wp:block-2 -->
  </div>

</div>

<!-- /wp:setka-editor/setka-editor -->
```

One of the problems is that we don't know where to place `<InnerBlocks />` component because Gutenberg blocks could be possibly placed anywhere inside our editor's content multiple times.

So we have two questions:
1. Is such kind of integration possible in general?
2. If yes, which Gutenberg API should help us to implement it?

Any thoughts and advices, pros and cons will be highly appreciated.

kmukku commented:
Hi, first of all I'm impressed of the whole Gutenberg project and efforts you guys have made to improve the editing experience in Wordpress. Good job! 

I have been lately working mostly with other CMS, which has had this sort of rendering engine for a long time. In previous versions it used to have some sort of column block like you have here, for nesting blocks, but it was removed as it was not necessary after all. I made a demo some time a go to demonstrate how to use CSS frameworks (same applies to CSS Grid) in order to manage the block rendering in UI. You can see it here: 

http://jsfiddle.net/kmukku/zz83dx0a/

Nothing fancy, but it gives the idea how to avoid use of column blocks. I'm sure this could be implemented pretty easily into Gutenberg too.
